Stream data to export file

DanielSpindler83 opened the pull request:
Please note this PR uses a branch based off the branch unit-testing also used in https://github.com/TurnerSoftware/Pingalot/pull/15
I reworked the pingsession.cs to be stateful - it is only passed a single ping request.
Removed the need for a list of PingRequests.
Moved the PingRequestExportModel class into Pingalot.Core
Added Export File to the PingRequestOptions - maybe not the best idea but was the easiest to get the file details passed through to the PingRequestAgent.
Added the File Export function into PingRequestAgent - export a record at a time as it occurs.
Catered for an error with the export file path - if error then create file local to the exe with a generic name plus datetime.
Added a final ping stats method to pingstats to just pass in the end time.
The memory is now constant for the life of the program as we don't keep a list of ping requests.
Updated existing unit-tests to ensure all passing.

DanielSpindler83 committed:
Started on two unit test methods for PingSession

DanielSpindler83 committed:
Refactor, cleanup comments. Add assert for TestPingStatisticsDivideByZero()

DanielSpindler83 committed:
merge in dividebyzero fix

DanielSpindler83 committed:
add asserts to main stats test and resolve some decimal rounding issues

DanielSpindler83 committed:
add in the divide by zero fix for AverageRoundTrip

DanielSpindler83 committed:
merge main into unit-testing - resolve conflicts manually.

DanielSpindler83 committed:
Apply code style changes already done elsewhere.

DanielSpindler83 committed:
redesign PingSession to be stateful. No longer requires a list of PingRequests, just a single add each call.

DanielSpindler83 committed:
Add command line arg ExportFile to PingRequestOptions - so we can use it in PingRequestAgent.

DanielSpindler83 committed:
remove old export code from program.cs

DanielSpindler83 committed:
redesign PingRequestAgent to use new stateful PingSession.

DanielSpindler83 committed:
redesign unit tests to cater for new PingSession stateful design.

DanielSpindler83 committed:
change TotalRoundtrip to a double so we can round AverageRoundTrip to two decimal places - fix unit test error.

DanielSpindler83 committed:
remove unused pingRequests List

DanielSpindler83 committed:
Add export file components to PingRequestAgent.

DanielSpindler83 committed:
closes TurnerSoftware/Pingalot#6

Turnerj commented on tests/Pingalot.Core.Tests/PingSessionTests.cs:
```diff
@@ -17,10 +17,10 @@ public void TestPingStatisticsDivideByZero()
 
 			var startTime = new DateTime(2022, 5, 19, 14, 30, 0);
-			var duration = TimeSpan.FromSeconds(30);
+			TimeSpan duration = new System.TimeSpan(0, 0, 0, 30);
```
Use `var` and drop `System.`

Turnerj commented on tests/Pingalot.Core.Tests/PingSessionTests.cs:
```diff
@@ -32,11 +32,86 @@ public void TestPingStatisticsDivideByZero()
+					HasMatchingBuffer = true,
+					RequestTime = DateTime.Now
+				};
+
+				testPingSession.AddSinglePingResult(duration, pingRequest);
+			}
+
+			// Create 2 successful pings 5ms RT 
+			for (int i = 0; i < 3; i++)
+			{
+				var pingRequest = new PingRequest
+				{
+					Address = IPAddress.Loopback,
+					Status = IPStatus.Success,
+					RoundtripTime = 5,
+					TimeToLive = 58,
+					BufferLength = 32,
+					HasMatchingBuffer = true,
+					RequestTime = DateTime.Now
+				};
+
+				testPingSession.AddSinglePingResult(duration, pingRequest);
+			}
+
+			// Create 2 TimedOut pings
+			for (int i = 0; i < 2; i++)
+			{
+				var pingRequest = new PingRequest
+				{
+					Address = IPAddress.Loopback,
+					Status = IPStatus.TimedOut,
+					RoundtripTime = 0,
+					TimeToLive = 0,
+					BufferLength = 0,
+					HasMatchingBuffer = false,
+					RequestTime = DateTime.Now
+				};
+
+				testPingSession.AddSinglePingResult(duration, pingRequest);
+			}
```
Rather than do these loops, I'd rather create a private static function in the test class that can create the `PingRequest` for us. Something that takes everything _but_ the address (we can default to `IPAddress.Loopback` unless we want to test IPv6 but that is kinda outside of scope for these tests).

Something like: `private static PingRequest CreatePingRequest(IPStatus status, double roundtripTime = 0, ...)` etc.

Then you'd just call this a few times in a row:
```csharp
testPingSession.AddSinglePingResult(duration, CreatePingResult(/* the parameters */));
testPingSession.AddSinglePingResult(duration, CreatePingResult(/* the parameters */));
...
```

I'd still avoid passing in something like `DateTime.Now` as it creates variability in test results (though depends on what we're testing).

Turnerj commented on src/Pingalot.Core/PingRequestAgent.cs:
```diff
@@ -13,20 +16,37 @@ public class PingRequestAgent
-			var buffer = CreateBuffer(options.BufferSize);
+			if (ExportFile != null)
+			{
+				try
+				{
+					SetupExportFile(ExportFile);
+				}
+				catch
+				{
+					// something went wrong with using the provided export file path\filename - so lets setup one local to exe
+					var fileNameDate = DateTime.Now.ToString("yyyy-MM-dd__HH-mm-ss");
+					ExportFile = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location) + "\\results_" + fileNameDate + ".csv";
+					SetupExportFile(ExportFile);
+				}
```
I wouldn't worry about `try`/`catch` here - if we can't setup the export file, just let it fail.

DanielSpindler83 replied:
@Turnerj  fair enough. 
I am thinking we could accept Export command line parameter without a path and default the path to where the exe runs and the filename with the date time created(same as the catch code above)? Thoughts? I personally liked using a default like that.
Side question - should we do any input validation\sanitation on our command line input? e.g. validate a path is a valid path or an IP is an IP? Maybe something we can add later?

Turnerj replied:
> I am thinking we could accept Export command line parameter without a path and default the path to where the exe runs and the filename with the date time created(same as the catch code above)? Thoughts? I personally liked using a default like that.

Yeah, that sounds good.

> Side question - should we do any input validation\sanitation on our command line input? e.g. validate a path is a valid path or an IP is an IP? Maybe something we can add later?

Yes and Yes :)

Turnerj commented on src/Pingalot.Core/PingRequestAgent.cs:
```diff
@@ -13,20 +16,37 @@ public class PingRequestAgent
 
 		public async Task<PingSession> StartAsync(PingRequestOptions options, CancellationToken cancellationToken)
 		{
+			var ExportFile = options.ExportFile; // https://github.com/TurnerSoftware/Pingalot/issues/6
```
Don't need to add comments into the code linking back to the issue

Turnerj commented on src/Pingalot.Core/PingRequestAgent.cs:
```diff
@@ -95,5 +120,31 @@ private static bool CheckBuffer(byte[] expected, byte[] actual)
 			return true;
 		}
+
+		private void SetupExportFile(string exportFile)
+		{
+			// open file and write out the csv file headers - just once
+			// we use append as file may already exist - thats ok still write to it
+			using (var stream = File.Open(exportFile, FileMode.Append))
+			using (var writer = new StreamWriter(stream))
+			using (var csv = new CsvWriter(writer, CultureInfo.InvariantCulture))
+			{
+				csv.WriteHeader<PingRequestExportModel>();
+				csv.NextRecord();
+			}
+		}
+
+		private void WriteRecordToExportFile(string exportFile, PingRequest pingRequest)
+		{
+			// write a single pingrequest record to export file
+			using (var stream = File.Open(exportFile, FileMode.Append))
+			using (var writer = new StreamWriter(stream))
+			using (var csv = new CsvWriter(writer, CultureInfo.InvariantCulture))
+			{
+				var singleExportablePingResult = new PingRequestExportModel(pingRequest);
+				csv.WriteRecord(singleExportablePingResult);
+				csv.NextRecord();
+			}
+		}
```
While this approach works - opening and closing the file per request - it would be better to restructure the code to use a single stream/writer/csvwriter across the entire time. Not sure how best to do that off the top of my head right now but feel free to experiment with techniques to do so.

DanielSpindler83 replied:
@Turnerj I have had a play with this idea and I was able to get it to work, but, is it possible to have a single stream and be able to access the file externally with another program, on the file system? Across long ping sessions, I have found it useful to access the exported file in another app, like Notepad++ or BareTail, in order to see the detail of individual pings. This is possible when we open and close the file for each record. With a single stream I have tried flushing to disk and still the file is locked until the stream is closed.

DanielSpindler83 replied:
maybe another idea is to use a logger to push each ping request to disk....

Turnerj replied:
Had a quick look, we can set the `FileShare` for a `FileStream`: https://docs.microsoft.com/en-us/dotnet/api/system.io.fileshare?view=net-6.0

This will allow us to have an exclusive write while other programs can read. We can probably leverage that here.

DanielSpindler83 replied:
> Had a quick look, we can set the `FileShare` for a `FileStream`: https://docs.microsoft.com/en-us/dotnet/api/system.io.fileshare?view=net-6.0
> 
> This will allow us to have an exclusive write while other programs can read. We can probably leverage that here.

This is EXACTLY what i was looking for and my google fu was lacking..... :-( 
I have tested this and it works perfectly. Thanks
`using (var stream = File.Open(options.ExportFile, FileMode.Append,FileAccess.Write,FileShare.Read))`

WIP

DanielSpindler83 committed:
create a TestPingResult method. https://github.com/TurnerSoftware/Pingalot/pull/16#discussion_r882543254

DanielSpindler83 committed:
remove for loops creating test pingrequests.

DanielSpindler83 committed:
missed one test ping for  ConfirmPingSessionStatistics. Added in.

DanielSpindler83 committed:
remove uneccesary link

DanielSpindler83 committed:
experiment with single write file stream

Turnerj commented:
Just FYI - I haven't forgotten about this, just balancing a lot of plates at the moment. Will get back to this sometime this week. 🙂

DanielSpindler83 commented:
np, i understand, I got a bit busy myself too. Thanks for the heads up.

On Sun, Jun 26, 2022 at 12:07 PM James Turner ***@***.***>
wrote:

> Just FYI - I haven't forgotten about this, just balancing a lot of plates
> at the moment. Will get back to this sometime this week. 🙂
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/TurnerSoftware/Pingalot/pull/16#issuecomment-1166397526>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHGYXCPVHWDAAHIFUSUZUXDVQ63MVANCNFSM5WX3AJIQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

DanielSpindler83 committed:
add new -e flag for default export, refactor main perform pings loop, to support single write file stream. Add stream flag to allow file share read access.

DanielSpindler83 commented:
Hi @Turnerj  , here is my latest version of single file stream writer.
Whilst this is working, I have a dirty feeling this is far from elegant.

I have added a new command line switch (-e) as a flag to use a default export file path and name alongside the exe.
If an export path is provided, with or without -e, then the provided path is used.
To be brutally honest, i spent a LONG time trying to work out the whole export \ default thingy.

I refactored the main workload into a "PerformPings" method so I could conditionally wrap the single stream around it.
I have reservations about how this is done - i pass in a null csv writer for when we don't have an export file.....more than likely a better way.
If we don't get provided a path at all, then we run PerformPings without the using file steam wrappers.

Go nuts and poke some holes in this , as I'm sure its terrible in more ways than one.
I appreciate your time.

Thanks

Turnerj commented:
I think we can flip the code around a bit - rather than trying to pass the exporting logic in, we instead make it listen on the same event we are pushing data out by. Basically when we know we want to export, we create a new instance of some exporter class that manages the file access and have it listen on our `PingCompleted` event. Every time the event is raised, we write the record to disk.

DanielSpindler83 commented:
> I think we can flip the code around a bit - rather than trying to pass the exporting logic in, we instead make it listen on the same event we are pushing data out by. Basically when we know we want to export, we create a new instance of some exporter class that manages the file access and have it listen on our `PingCompleted` event. Every time the event is raised, we write the record to disk.

I can have a go at that.....

DanielSpindler83 committed:
setup and test FileExporter event

DanielSpindler83 committed:
add new -e flag for default export and refactor pingarguments

DanielSpindler83 commented:
> I think we can flip the code around a bit - rather than trying to pass the exporting logic in, we instead make it listen on the same event we are pushing data out by. Basically when we know we want to export, we create a new instance of some exporter class that manages the file access and have it listen on our `PingCompleted` event. Every time the event is raised, we write the record to disk.

So I have a new file exporter class and I have setup a method to listen to the `PingCompleted` event. Im thinking we can no longer use a single set of using statements as the event leaves that code block, then comes back and the filestream objects get cleaned up - I get  "object reference not set to an instance of an object"...exception. I am thinking we dont use the using statements and have the fileexporter dispose of the streams once the instance of the fileexporter is no longer used - with the file steams as member of the file exporter....
Is this the correct approach?

Turnerj commented:
> I am thinking we dont use the using statements and have the fileexporter dispose of the streams once the instance of the fileexporter is no longer used - with the file steams as member of the file exporter....

Yeah, pretty much that. We make the file exporter disposable (with a using statement etc) and make that dispose of our steam when it disposes.

DanielSpindler83 committed:
file exporter working with event. No using and no dispose.

DanielSpindler83 committed:
tested and working export

DanielSpindler83 committed:
merge in export event branch

DanielSpindler83 committed:
remove unused export methods

DanielSpindler83 committed:
bug fix missing parameter

DanielSpindler83 committed:
bug fix

DanielSpindler83 committed:
removed commented out code

DanielSpindler83 commented:
@Turnerj working version using an event to conduct the export.
I did experiment with a few different versions, however i feel i make things overly complex and borderline unreadable.
If i come back to it later on and I cant read my own code then its poor and i discard it.
Hopefully this is nice and simple.
Couldn't work out how to wrap my new PingRequestFileExporter in a using statement.

DanielSpindler83 committed:
added filexporter to traditional ping and tested

Turnerj commented on src/Pingalot/Layouts/ModernPing.cs:
```diff
@@ -13,6 +13,13 @@ public static async Task<PingSession> StartAsync(PingRequestOptions options)
 			var pingRequestAgent = new PingRequestAgent();
 			var cancellationTokenSource = new CancellationTokenSource();
 
+			if (options.ExportFileFullPath != null)
+			{
+				var pingRequestFileExporter = new PingRequestFileExporter(options.ExportFileFullPath);
+				pingRequestAgent.PingCompleted += pingRequestFileExporter.ExportSingleResultToFile;
+
+			}
+
```
OK, I see why you had trouble trying to dispose of it. A quick-and-dirty workaround would be to declare `PingRequestFileExporter? pingRequestFileExporter = null;` before the if-statement and then wrap the rest of the code in a try/finally. Then in the finally, we go `pingRequestFileExporter?.Dispose();`.

For a more complete solution, it might be a more ground-up redesign of how the classes interact with each other so we don't even need the file exporter within the specific visual implementations.

Turnerj replied:
What I mean by the last part is that some of the difficulty is really just how the classes call each other. Ideal world would have something like each "layout" just register events rather than handle the life cycle itself. The ping session is kinda redundant now as we don't do anything with it. Remove the replication from the cancel key press etc logic in each layout.

If I were to name what I'm thinking, something like "pipelines". Each thing (a layout, exporting, etc) is just an item on the pipeline. I guess kinda like how the middleware works for ASP.NET Core but more around events than just calling a `next();` action.
